src/main/java/seedu/address/logic/commands/TagCommand.java

```diff
@@ -23,59 +24,95 @@ public class TagCommand extends Command {
-     * @param tag to tag the person with
+     * @param targetIndexes of the person in the filtered person list to tag
+     * @param tags set of tags to tag the person with
      */
-    public TagCommand(Index targetIndex, Tag tag) {
-        requireNonNull(targetIndex);
-        requireNonNull(tag);
+    public TagCommand(List<Index> targetIndexes, Set<Tag> tags) {
+        requireNonNull(targetIndexes);
+        requireNonNull(tags);
 
-        this.targetIndex = targetIndex;
-        this.tag = tag;
+        this.targetIndexes = targetIndexes;
+        this.tags = tags;
     }
 
     @Override
     public CommandResult execute(Model model) throws CommandException {
         requireNonNull(model);
         List<Person> lastShownList = model.getFilteredPersonList();
 
-        if (targetIndex.getZeroBased() >= lastShownList.size()) {
-            throw new CommandException(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
+        Set<Tag> missingTags = new HashSet<>();
+        Set<Tag> duplicateTags = new HashSet<>();
+        StringBuilder successMessage = new StringBuilder();
+        StringBuilder finalMessage = new StringBuilder();
+
+        for (Index targetIndex : targetIndexes) {
+            if (targetIndex.getZeroBased() >= lastShownList.size()) {
+                throw new CommandException(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
+            }
+            Person personToTag = lastShownList.get(targetIndex.getZeroBased());
+            Set<Tag> newTags = new HashSet<>(personToTag.getTags());
+            boolean updated = false;
+
+            for (Tag tag : tags) {
+                if (!model.hasTag(tag)) {
+                    missingTags.add(tag);
+                } else if (newTags.contains(tag)) {
+                    duplicateTags.add(tag);
+                } else {
+                    newTags.add(tag);
+                    updated = true;
+                }
+            }
+
+            if (updated) {
+                Person updatedPerson = new Person(personToTag.getName(), personToTag.getPhone(),
+                        personToTag.getEmail(), personToTag.getRsvpStatus(), newTags);
+                model.setPerson(personToTag, updatedPerson);
+                if (successMessage.length() > 0) {
+                    successMessage.append("\n");
+                }
+                successMessage.append(Messages.format(updatedPerson));
+            }
         }
-
-        Person personToTag = lastShownList.get(targetIndex.getZeroBased());
-
-        if (!model.hasTag(tag)) {
-            throw new CommandException(tag + " " + MESSAGE_TAG_NOT_CREATED);
+        if (successMessage.length() > 0) {
+            finalMessage.append(MESSAGE_TAG_PERSON_SUCCESS).append(successMessage);
         }
-
-        if (personToTag.getTags().contains(tag)) {
-            throw new CommandException(MESSAGE_DUPLICATE_TAG);
+        if (!missingTags.isEmpty() || !duplicateTags.isEmpty()) {
```

Comment:
Similarly, can the handling of messages be extracted out so that the method focuses mainly on the logic of tagging?